### PR TITLE
I added a new class pollingWorker in a new thread that prints values of parameters to the terminal at rates specified in the config file

### PR DESCRIPTION
**Pull request by JonathanNolan3**
new file created "pollingWorker.py".
[serverConfigExample.txt](https://github.com/toolsforexperiments/instrumentserver/files/14729180/serverConfigExample.txt)
The example config file demonstrates how to use the worker. Adding the pollingRates field will create a new Qtimer for each parameter with the specified interval in the config file (in seconds). Timers will only be created for existing parameters when the server is started. Adding new parameters in the parameter manager will not create new timers. The worker runs in a new Qthread "pollingThread" created in apps.py. The thread is stopped when the server is stopped in core.py.

### CODE REVIEW

File: instrumentserver/apps.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from .client import Client
 from .gui import widgetDialog, widgetMainWindow
 from .gui.instruments import ParameterManagerGui
-
+from .server.pollingWorker import pollingWorker
```
Comment (marcosfrenkel) [inline instrumentserver/apps.py:22]: add a white-space between imports and code

File: instrumentserver/apps.py [hunk 2]
```diff
@@ -83,6 +90,9 @@ def serverScript() -> None:
         tempFile.close()
         Path(stationConfig).unlink(missing_ok=True)
 
+def quitPollingThread():
```
Comment (marcosfrenkel) [inline instrumentserver/apps.py:93]: add a '_' at the beginning of the function name. This indicates that this is a "private" method and users should not use this function. Should look like `_quitPollingThread`

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

File: instrumentserver/server/core.py [hunk 3]
```diff
@@ -234,6 +234,9 @@ def startServer(self) -> None:
 
             self.messageReceived.emit(str(message), response_log)
 
+        from ..apps import quitPollingThread
```
[the remark this answers is not in the document]
Comment (JonathanNolan3) [reply instrumentserver/server/core.py:237]: Note: Causes circular import if placed at the beginning of the file

File: instrumentserver/config.py [hunk 4]
```diff
@@ -63,6 +64,12 @@ def loadConfig(configPath: str):
         else:
             guiConfig[instrumentName] = GUIFIELD
 
+        if 'pollingRate' in configDict:
+            pollingRateList = configDict.pop('pollingRate')
+            if pollingRateList != None:
```
Comment (marcosfrenkel) [inline instrumentserver/config.py:69]: change the != for `is not`. When dealing with `None` that is a better way of doing that comparison in python

File: instrumentserver/apps.py [hunk 5]
```diff
@@ -58,10 +59,17 @@ def serverScript() -> None:
     # Load and process the config file if any.
     configPath = args.config
 
-    stationConfig, serverConfig, guiConfig, tempFile = None, None, None, None
+    stationConfig, serverConfig, guiConfig, tempFile, pollingRates = None, None, None, None, None
     if configPath != '':
         # Separates the corresponding settings into the 4 necessary parts
```
Comment (marcosfrenkel) [inline instrumentserver/apps.py:64]: into the 5 necessary parts

File: instrumentserver/server/pollingWorker.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    # Creates a qtimer for each param in the dict with the interval specified
+    def run(self):
+        timers = []
+
+        # Deletes param from dict if it does not exist
+        delList = []
+        for param in self.pollingRatesDict:
+            if param not in self.cli.getParamDict(param.split(".")[0]):
+                logger.warn(f"Parameter {param} does not exist")
```
Comment (marcosfrenkel) [inline instrumentserver/server/pollingWorker.py:29]: The warn method is deprecated, use warning instead

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

**Comment by marcosfrenkel** [comment]
I have added a few more cosmetic changes that I found when I was looking over the code again. 

I still think there should be a way of avoiding that import in the middle of the code. Let me see if I can figure some workaround

File: instrumentserver/server/core.py [hunk 7]
```diff
@@ -233,7 +233,10 @@ def startServer(self) -> None:
             send(socket, response_to_client)
 
             self.messageReceived.emit(str(message), response_log)
-
+        # Importing at the top of the file causes a circular import
+        from ..apps import _quitPollingThread
+        _quitPollingThread()
```
Comment (marcosfrenkel) [inline instrumentserver/server/core.py]: This should be removed right? it is not needed anymore now that we are passing the thread as an argument?

File: instrumentserver/apps.py [hunk 8]
```diff
@@ -58,10 +59,17 @@ def serverScript() -> None:
...
-        stationConfig, serverConfig, guiConfig, tempFile = loadConfig(configPath)
+        stationConfig, serverConfig, guiConfig, tempFile, pollingRates = loadConfig(configPath)
+
+    serverScript.pollingThread = QtCore.QThread()
```
Comment (marcosfrenkel) [inline instrumentserver/apps.py]: We should probably have this only start if the user passes a polling rate. But we can have that as an issue for now if its too hard

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

File: instrumentserver/apps.py [hunk 9]
```diff
@@ -58,25 +59,34 @@ def serverScript() -> None:
...
 
-    stationConfig, serverConfig, guiConfig, tempFile = None, None, None, None
+    stationConfig, serverConfig, guiConfig, tempFile, pollingRates = None, None, None, None, None
     if configPath != '':
-        # Separates the corresponding settings into the 4 necessary parts
-        stationConfig, serverConfig, guiConfig, tempFile = loadConfig(configPath)
+        # Separates the corresponding settings into the 5 necessary parts
+        stationConfig, serverConfig, guiConfig, tempFile, pollingRates = loadConfig(configPath)
+
+    pollingThread = QtCore.QThread()
```
Comment (marcosfrenkel) [inline instrumentserver/apps.py:67]: Add an if that thins only happens if pollinRates is not None, no need to start a new polling thread if no polling is accepted.

File: instrumentserver/server/core.py [hunk 10]
```diff
@@ -97,6 +97,7 @@ def __init__(self,
                  initScript: Optional[str] = None,
                  serverConfig: Optional[Dict[str, Any]] = None,
                  stationConfig: Optional[str] = None,
+                 pthread: Any = None
```
Comment (marcosfrenkel) [inline instrumentserver/server/core.py:100]: rename it pollingThread

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]
last things and this is ready for merging

**Pull request merged**